rfcs/61-add-object-mode-check-for-operations.md

```diff
@@ -0,0 +1,251 @@
+- Pager
+  - WritePage
+
+## Proposal
+
+So I propose that we should add mode check in specific operation, return `ObjectModeInvalidError` if mode not meet, and the check should be generated.
+
+To generate object mode check in different functions, we will follow these steps (take `Multiparter.WriteMultipart` as an example):
+
+1. Add a field named `object_mode` in `operations.toml`.
+```toml
+[multiparter.op.write_multipart]
+description = "will write content to a multipart."
+params = ["o", "r", "size", "index"]
+results = ["n", "part"]
+object_mode = "part"
+```
+
+2. Add `ObjectMode` field parsing in `parse.go` and `spec.go`.
+```go
+// tomlOperation in parse.go
+type tomlOperation struct {
+    Description string   `toml:"description"`
+    Params      []string `toml:"params"`
+    Pairs       []string `toml:"pairs"`
+    Results     []string `toml:"results"`
+    ObjectMode  string   `toml:"object_mode"` // add ObjectMode
+    Local       bool     `toml:"local"`
+}
+
+// Operation in spec.go
+type Operation struct {
+    Name        string
+    Description string
+    Params      []string
+    Pairs       []string
+    Results     []string
+    ObjectMode  string // add ObjectMode
+    Local       bool
+}
+```
+
+3. Modify relevant struct and parse func in [go-storage]
+
+```go
+// cmd/definitions/type.go
+type Operation struct {
+    Name        string
+    Description string
+    Pairs       []string
+    Params      Fields
+    Results     Fields
+    ObjectMode  string // add ObjectMode
+    Local       bool
+}
+```
+```go
+// cmd/definitions/type.go
+func NewOperation(v specs.Operation, fields map[string]*Field) *Operation {
+    op := &Operation{
+        Name:        v.Name,
+        Local:       v.Local,
+        ObjectMode:  v.ObjectMode, // pass ObjectMode from specs.Operation
+        Description: formatDescription("", v.Description),
+    }
+    ...
+}
+```
+```go
+// cmd/definitions/type.go
+// HasObject will check whether we have *Object here.
+func (f Fields) HasObject() bool {
+    for _, v := range f {
+        if v.ftype == "Object" || v.ftype == "*Object" {
+            return true
+        }
+    }
+    return false
+}
+
+// ObjectParamName returns Object's param name.
+func (f Fields) ObjectParamName() string {
+    for _, v := range f {
+        if v.ftype == "Object" || v.ftype == "*Object" {
+            return v.Name
+        }
+    }
+    return "o"
+}
+```
+
+4. Add templates for code generating in [go-storage]
+
+```go
+// cmd/definitions/tmpl/service.tmpl
+// {{ $fnk }}WithContext {{ $fn.Description }}
+func (s *{{$pn}}) {{ $fnk }}WithContext(ctx context.Context, {{$fn.Params.String}}) ({{$fn.Results.String}}) {
+    defer func (){
+        {{- $path := $fn.Params.PathCaller }}
+        {{- if and (eq $path "") (eq $pn "Service") }}
+            {{ $path = ",\"\"" }}
+        {{- end }}
+        err = s.formatError("{{$fn.Name}}", err {{ $path }} )
+    }()
+    
+    {{- if $fn.Params.HasObject }}
```

Comment:
Maybe we can simplify the check here? All operations that have `ObjectMode` must have `o *Object`, so we can use them directly.